docs/Boxes.md

```diff
@@ -0,0 +1,140 @@
+the `Option<int>` type) getting that returned value.So this operates something like exceptions.
+See the [reference manual](TODO)) for more on the similarities and differences with exceptions.
+
+### Outcome
+
+`Outcome` is a variation on these boxed types in which we mostly just care whether the operations succeeded or failed.
+It has no place to carry a value, just information about the error. 
+
+This datatype is most useful when used with methods with multiple return values.
+Then the first return value can be an `Outcome` to indicate success or failure, with the other outputs returning 
+appropriate values in the success situation.
+
+For example, if we have this method
+```dafny
+method FindAllMatches<T(==)(s: seq<T>, value: T) returns (status: Outcome<string>, matches: seq<int>)
+```
+we can call it as
+```dafny
+var matches :- FindAllMatches(s, value);
+```
+
+Notice that there is no left-hand-side for the first out-parameter. It does not carry a value: if the value is a Fail, the control
+flow will abruptly return; if is is a Pass, the first out-parameter is then discarded and the second is assigned to the remaining LHS.
+An Outcome serves as an automatically-checked erorr mechanism.
+
+
+### Result
+
+A `Result` carries both a success value and failure information, or two separate, specifiable types. Its use is then very similar to `Option`.
+
+### Combining different FC-types in expressions
+
+The example given for `Option` above is legal Dafny code because the output of the caller method `m` has the same type
+(`Option<int>`) as the output of the callee function `Find`. But what if it does not? You may be writing code that generally uses, say, `Result`,
+but happens to be using code supplied by someone else that returns an `Outcome`.
+
+When values of these types are used in expressions, the library types offer means to convert among them: each type has instance functions that
+convert to values of the other types. For example, we can rewrite the example above this way:
+
+```dafny
+import opened Dafny.Boxes
+
+function Find<T(==)>(s: seq<T>, k: int, value: T): (r: Option<int>)
+  requires 0 <= k <= |s|
+  ensures  r.Some? ==> k <= r.Extract() < |s|
+  ensures  r.Some? ==> s[r.Extract()] == value;
+  ensures  r.None? ==> forall i | k <= i < |s| :: s[i] != value
+  decreases |s| - k
+{
+  if k >= |s| then None else if s[k] == value then Some(k) else Find(s, k+1, value)
+}
+
+method m(s: seq<int>) returns (r: Result<int,string>) {
+  var value: int := *;
+  // do various calculations
+  var index: int :- Find(s, 0, value).ToResult("not found");
+  // more calculations
+  return Success(value);
+}
+```
+
+Here we used the `ToResult` function of `Option` to convert the result value to a `Result`. There are similar `ToOutcome` and `ToOption` functions in each type.
+
+But what if we need to convert to some custom FC-type? Each datatype also has a `Map` function that takes a function as an argument, which function is a converter from
```

Comment:
```suggestion
But what if we need to convert to some custom FC-type? Each datatype comes with a `Map` function, that can be given as an argument a function that converts from
```

Reply:
Not sure which `Map` function you are referring to, but it is not the usual one, and there is no alternative one in the current library.

Here you define an (anonymous) function `f: Option<int> -> Result<int, string>`, and this function can be applied to the output of type `Option<int>` of the function `Find`. I see no need to use `f` as an argument to a `Map` function, if you set up things in this way. Unless you define `Map(f) := f(this)`? If that's what you have in mind, I propose not to use the name 'Map', as this is often used to indicate the lifting of `f: A -> B` to e.g. `Option<A> -> Option<B>`.

Reply:
Do you have an alternate, concise, intuitive name?

Map(f) is indeed f(this) -- I added it because I wanted to provide the alternative of a streaming syntax x.Map(f) as well as f(x), for which no additional syntax is needed.